### PR TITLE
Refactor loop check for inout modification into a recursive function

### PR DESCRIPTION
**Pull request by SSoulaimane**
Fragment from #9282

### CODE REVIEW

File: src/dmd/expressionsem.d [hunk 1]
```diff
@@ -1885,27 +1885,31 @@ private bool functionParameters(const ref Loc loc, Scope* sc,
              *   S().bar() = 1;      // bad!
              * }
              */
-            Dsymbol s = null;
-            if (fd.isThis() || fd.isNested())
-                s = fd.toParent2();
-            for (; s; s = s.toParent2())
+            bool checkEclosingWild(Dsymbol s)
```
Comment (JinShil) [inline src/dmd/expressionsem.d]: checkEclosing --> checkEnclosing  Missing an 'n'.

Comment (SSoulaimane) [reply src/dmd/expressionsem.d]: If we could teach the compiler English grammar. Thanks!

File: src/dmd/expressionsem.d [hunk 2]
```diff
@@ -1885,27 +1885,31 @@ private bool functionParameters(const ref Loc loc, Scope* sc,
              *   S().bar() = 1;      // bad!
              * }
              */
-            Dsymbol s = null;
-            if (fd.isThis() || fd.isNested())
-                s = fd.toParent2();
-            for (; s; s = s.toParent2())
+            bool checkEnclosingWild(Dsymbol s)
```
Comment (Geod24) [inline src/dmd/expressionsem.d]: Looks like both nested functions can be static ?

**Comment by JinShil** [comment]
Forgive my ignorance, but please provide some rationale for how this improves things.  It's trading a simple for loop for recursion.  Recursion has created a serious problem in the past resulting in a stack overflow in this code.  See [Issue 18026](https://issues.dlang.org/show_bug.cgi?id=18026).

File: src/dmd/expressionsem.d (hunk 2 above)
Comment (SSoulaimane) [reply src/dmd/expressionsem.d]: if they didn't call `errorInout()`.

**Comment by SSoulaimane** [comment]
> Forgive my ignorance, but please provide some rationale for how this improves things. It's trading a simple for loop for recursion. Recursion has created serious problems in the past resulting in stack overflows. See [Issue 18026](https://issues.dlang.org/show_bug.cgi?id=18026).

My bad I should have explained the connection to #9282.
This PR has zero benefit on it's own but at https://github.com/dlang/dmd/pull/9282#discussion_r275437400 where this recursion is used to walk a tree instead and it was suggested there to make it into a separate refactoring to make #9282 smaller in size.

**Comment by JinShil** [comment]
> My bad I should have explained the connection to #9282.

No worries.  I see that you need to create a function, but is recursion necessary?  Can you maintain the loop structure, but just warp it in a function?  Or is the recursion providing some benefit I'm not seeing?

**Comment by SSoulaimane** [comment]
> No worries. I see that you need to create a function, but is recursion necessary? Can you maintain the loop structure, but just warp it in a function? Or is the recursion providing some benefit I'm not seeing?

see here https://github.com/dlang/dmd/pull/9282/commits/2a20411ba43c1f30f00f86c15e1a6d6783420ac4#diff-0c1aa1ff99402eb87e11cc3fa1b80da0R1918 if I kept the loop I would need to use a stack there because I am crawling a binary tree of parents.

**Review by JinShil — APPROVED** [review]

**Review by thewilsonator — APPROVED** [review]

**Pull request merged**